SplunkRumWorkspace/SplunkRum/SplunkRum/ScreenFrames.swift

```diff
@@ -0,0 +1,104 @@
+//
+/*
+Copyright 2021 Splunk Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import Foundation
+import QuartzCore
+import UIKit
+
+class ScreenFrames: NSObject {
+
+    private var displayLink: CADisplayLink?
+
+    private var slowFrameThreshold: CFTimeInterval = 1.0 / 59.0
+    private var frozenFrameThreshold: CFTimeInterval = 700.0 / 1000.0
+
+    private var currentIteration: Int = 0
+    private var startedTime: CFTimeInterval = CACurrentMediaTime()
+
+    private var slowCount: Int = 0
+    private var frozenCount: Int = 0
+    private var isFirstIteration: Bool = true
+    private var previousTimestamp: CFAbsoluteTime = CACurrentMediaTime()
+
+    func startTracking() {
+
+        stopTracking() /// make sure to stop a previous running display link
+        let displayLink = CADisplayLink(target: self, selector: #selector(displayLinkCallback))
+        displayLink.add(to: .main, forMode: .common)
+        self.displayLink = displayLink
+    }
+
+    func stopTracking() {
+
+        displayLink?.invalidate()
+        displayLink = nil
+    }
+
+    @objc func displayLinkCallback(_ displayLink: CADisplayLink) {
+
+         let state = UIApplication.shared.applicationState
+         if state == .background || state == .inactive {
+            isFirstIteration = true
+         }
+         let currentTime: CFTimeInterval = CACurrentMediaTime()
+         if isFirstIteration {
+            previousTimestamp = displayLink.targetTimestamp
+            isFirstIteration = false
+         }
+         let duration = displayLink.timestamp - previousTimestamp
+         previousTimestamp = displayLink.timestamp
+         let elapsedTime = currentTime - startedTime
+         let iteration = Int(elapsedTime)
+         if currentIteration == iteration {
+
+            if duration > frozenFrameThreshold {
+                frozenCount += 1
+             } else if duration > slowFrameThreshold {
+                slowCount += 1
+             }
+
+         } else {
+
+             if slowCount > 0 {
+                 reportSlowframe(slowFrameCount: slowCount, name: "slowRenders")
+             }
+
+             if frozenCount > 0 {
+                 reportSlowframe(slowFrameCount: frozenCount, name: "frozenRenders")
+             } else {
+                 if duration > frozenFrameThreshold {
+                     reportSlowframe(slowFrameCount: 1, name: "frozenRenders")
+                 }
```

Comment:
The outer else branch is taken at the beginning of a new full second. Our goal was to bucket the counts to 1 second intervals. This line seems to report a frozen frame at the beginning of a new second. Now if there will be any additional frozen frames during the next second, there will be 2 spans generated instead of 1.

Reply:
@seemk, there will only be one span generated per second with this logic. You are correct that the previous span will be sent at the beginning of the new second. So if there are additional frozen frames generated during the next second that span will be sent in the subsequent iteration. Let me know if I am missing something here. We can take a short call to discuss this and conclude.